### PR TITLE
Generalize gamm Inverse relationship test , add Inverse relationship test for stableswap

### PR DESCRIPTION
**Pull request by mattverse**
Closes: [#1371](https://github.com/osmosis-labs/osmosis/issues/1371)

## What is the purpose of the change

Upon adding invariant test for inverse relationship on `CalculateAmountOutAndIn`(https://github.com/osmosis-labs/osmosis/blob/main/x/gamm/pool-models/balancer/amm_test.go#L89-L216) the goal of this PR is to generalize the test case, and also add the same test for stableswap using the generalized test case.

*Add a description of the overall background and high level changes that this PR introduces *

*(E.g.: This pull request improves documation of area A by adding ....*

## Testing and Verifying

No testings changed in `balancer` package, inverse relationship test added for `stableswap`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/amm_test.go:203]: I see that all this setup is duplicated with the balancer pool. Can we also move everything on lines 190-203 to the common test suite `CfmmCommonTestSuite`?

File: app/apptesting/test_suite.go
Comment (p0mvn) [inline app/apptesting/test_suite.go:54]: I don't think this is used anywhere. Let's remove it?

File: x/gamm/pool-models/balancer/amm_test.go
Comment (p0mvn) [inline x/gamm/pool-models/balancer/amm_test.go:15]: Should we instead merge `KeeperTestSuite` and `BalancerTestSuite` but keep the `BalancerTestSuite` name? I think all tests on a common abstraction layer should belong to the same suite and be named with the respect to the abstraction layer it is testing

File: x/gamm/pool-models/stableswap/util_test.go
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/util_test.go:27]: I think this function is being shared by both pool packages. Can we instead move it to the common package and reuse?

**Review by p0mvn — COMMENTED** [review]
Great work @xBalbinus @mattverse ! Left some comments, please take a look

File: x/gamm/pool-models/balancer/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:15]: wdym by "merging" `KeeperTestSuite` and `BalancerTestSuite`?

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/stableswap/amm_test.go:114]: yeah agreed that we don't need weights here! Per scaling factors, I was thinking that we should address them in a separate PR. Would you mind if we create an issue here for creating test cases for stable swap for now?

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/stableswap/amm_test.go:125]: Just curious, why so?

Comment (p0mvn) [reply x/gamm/pool-models/stableswap/amm_test.go:125]: Just to be aligned with the stabelswap abstraction to avoid confusion for new folks reading the tests. Functionally, there is no difference ofc. Let's not block the PR on this. Up to you if you'd like to change

File: x/gamm/pool-models/balancer/amm_test.go
Comment (p0mvn) [reply x/gamm/pool-models/balancer/amm_test.go:15]: Here's the `BalancerTestSuite`: https://github.com/osmosis-labs/osmosis/blob/9488499f200642666ee3fe5d102ab197ebbcafde/x/gamm/pool-models/balancer/amm_test.go#L15

In the same file we are using `KeeperTestSuite`: https://github.com/osmosis-labs/osmosis/blob/9488499f200642666ee3fe5d102ab197ebbcafde/x/gamm/pool-models/balancer/amm_test.go#L21

Which is defined in `suite_test.go`: https://github.com/osmosis-labs/osmosis/blob/ce13399bf5271ab5827e56f9093149cb6b84feb0/x/gamm/pool-models/balancer/suite_test.go#L12

I see two potential problems here:
1. There are 2 suites on the same layer of abstraction (`balancer`)
   a. `KeeperTestSuite`
   b. `BalancerTestSuite`
2.  `KeeperTestSuite` name doesn't make sense because this is not testing the keeper layer

So I suggest removing `KeeperTestSuite` and porting all its tests to the `BalancerTestSuite` so that we have one suite in the `balancer` package

Comment (alexanderbez) [reply x/gamm/pool-models/balancer/amm_test.go:15]: @p0mvn that makes sense to me as well

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/stableswap/amm_test.go:203]: For this comment, seems to me that moving line 190-203 as a separate method would cause an extra overhead, since there's not much logic here and the method would simply return what it got as a parameter, when we can just assign values directly. Which would look like this
```
func (suite *CfmmCommonTestSuite) preparePoolVariables(
      denomOut string,
      initialPoolOut int64,
      denomIn string, 
      initialPoolIn string,
      swapFee string,
      exitFee string)(poolLiquidityIn sdk.Coin, poolLiquidityOut sdk.Coin, swapFeeDec sdk.Dec, exitFeeDec sdk.Dec) {
}
```

File: app/apptesting/test_suite.go
Comment (mattverse) [reply app/apptesting/test_suite.go:54]: Added it since I thought we would use it one day :p Agreed to removing it for now, But we should definitely utilize the apptesting package better!

File: x/gamm/pool-models/balancer/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:15]: Oh yeah definitely makes sense! sgtm

Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:15]: The original reason I left two different Suites was because of the two following reasons.
1. Having ` BalancerTestSuite` embed both `apptesting.KeeperTestHelper`, test_helpers.CfmmCommonTestSuite` causes error since we can't embed two same fields in one struct, and `apptesting.KeeperTestHelper` and `test_helpers.CfmmCommonTestSuite` both embed suite.Suite as default. 

We have a couple options as a get around for this, this one I can think of right now is removing suite.Suite embedding from `KeeperTestHelper`and setting it as a field like this 
```
type KeeperTestHelper struct {
	ApptestingSuite suite.Suite
	App             *app.OsmosisApp
	Ctx             sdk.Context
	QueryHelper     *baseapp.QueryServiceTestHelper
	TestAccs        []sdk.AccAddress
}
```

What do you think?

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/stableswap/amm_test.go:114]: Thanks!

**Comment by stackman27** [comment]
Closes [#1349](https://github.com/osmosis-labs/osmosis/issues/1349)?

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (stackman27) [inline x/gamm/pool-models/stableswap/amm_test.go:110]: Hi @mattverse how do i test this function? This is what i get when i try to run the command

File: x/gamm/pool-models/balancer/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:15]: cool, leaving as it is!

File: x/gamm/pool-models/stableswap/amm_test.go
Comment (mattverse) [reply x/gamm/pool-models/stableswap/amm_test.go:110]: so this is a part of a test suite, not a direct go test. I'd recommend using make file

File: x/gamm/pool-models/stableswap/util_test.go
Comment (mattverse) [reply x/gamm/pool-models/stableswap/util_test.go]: done in the latest commit

**Comment by mattverse** [comment]
I didn't notice this PR was open for so long, just made the changes you've requested per last review can you review the updates? tysm!

File: x/gamm/pool-models/balancer/amm_test.go
Comment (alexanderbez) [inline x/gamm/pool-models/balancer/amm_test.go:93]:
```suggestion
func (suite *BalancerTestSuite) TestBalancerCalculateAmountOutAndIn_InverseRelationship(t *testing.T) {
```

File: x/gamm/pool-models/internal/test_helpers/test_helpers.go
Comment (alexanderbez) [inline x/gamm/pool-models/internal/test_helpers/test_helpers.go:17]: can we add a small godoc to what CFMM abbreviation is?

File: x/gamm/pool-models/internal/test_helpers/test_helpers.go
Comment (alexanderbez) [inline x/gamm/pool-models/internal/test_helpers/test_helpers.go:38]:
```suggestion
	swapFee sdk.Dec,
) {
```

**Review by alexanderbez — APPROVED** [review]

File: x/gamm/pool-models/internal/test_helpers/test_helpers.go
Comment (ValarDragon) [inline x/gamm/pool-models/internal/test_helpers/test_helpers.go:20]: Why aren't we just using apptesting here?

Comment (ValarDragon) [reply x/gamm/pool-models/internal/test_helpers/test_helpers.go:20]: We could also delete the CreateTestContext if we used apptesting

Comment (mattverse) [reply x/gamm/pool-models/internal/test_helpers/test_helpers.go:20]: It's because we have import cycle problems when we do that since the apptesting package imports gamm module.

**Review by ValarDragon — APPROVED** [review]

**Comment by ValarDragon** [comment]
uhh, could you take a stab at solving the merge conflict here @mattverse , then we can get it merged in?

**Comment by mattverse** [comment]
np! Thanks for getting this reviewed! Merging once final CI checks have passed upon approval

**Pull request merged**
